cibuildwheel/logger.py

```diff
@@ -0,0 +1,259 @@
+import codecs
+import os
+import re
+import sys
+import time
+from typing import Optional, Union
+
+DEFAULT_FOLD_PATTERN = ('{name}', '')
+FOLD_PATTERNS = {
+    'azure': ('##[group]{name}', '##[endgroup]'),
+    'travis': ('travis_fold:start:{identifier}\n{name}', 'travis_fold:end:{identifier}'),
+    'github': ('::group::{name}', '::endgroup::{name}'),
+}
+
+PLATFORM_IDENTIFIER_DESCIPTIONS = {
+    'manylinux_x86_64': 'manylinux x86_64',
+    'manylinux_i686': 'manylinux i686',
+    'manylinux_aarch64': 'manylinux aarch64',
+    'manylinux_ppc64le': 'manylinux ppc64le',
+    'manylinux_s390x': 'manylinux s390x',
+    'win32': 'Windows 32bit',
+    'win_amd64': 'Windows 64bit',
+    'macosx_x86_64': 'macOS x86_64',
+}
+
+
+class Logger:
+    fold_mode: str
+    colors_enabled: bool
+    unicode_enabled: bool
+    active_build_identifier: Optional[str] = None
+    build_start_time: Optional[float] = None
+    step_start_time: Optional[float] = None
+    active_fold_group_name: Optional[str] = None
+
+    def __init__(self):
+        if sys.platform == 'win32' and hasattr(sys.stdout, 'reconfigure'):
+            # the encoding on Windows can be a 1-byte charmap, but all CIs
+            # support utf8, so we hardcode that
+            sys.stdout.reconfigure(encoding='utf8')
+
+        self.unicode_enabled = file_supports_unicode(sys.stdout)
+
+        if 'AZURE_HTTP_USER_AGENT' in os.environ:
+            self.fold_mode = 'azure'
+            self.colors_enabled = True
+
+        elif 'GITHUB_ACTIONS' in os.environ:
+            self.fold_mode = 'github'
+            self.colors_enabled = True
+
+        elif 'TRAVIS' in os.environ:
+            self.fold_mode = 'travis'
+            self.colors_enabled = True
+
+        elif 'APPVEYOR' in os.environ:
+            self.fold_mode = 'disabled'
+            self.colors_enabled = True
+
+        else:
```

Comment:
Don't we already detect the CI systems somewhere else in `cibuildwheel`? Can we share this detection?

Reply:
good point, I'll look at pulling that into a util